shop.cpp

```diff
@@ -1,26 +1,77 @@
 #include "shop.h"
-
+#include <chrono>
 
 Shop::Shop(int mMaxChairs)
     :mMaxChairs(mMaxChairs)
-    ,mTheBarber(this)
-    ,mBarberSemaphore(0)
+    ,mBarberSemaphore(1)
     ,mCustomersSemaphore(0)
 {
+    mTheBarber = new Barber(this);
+}
+
+Shop::~Shop()
+{
+    delete mTheBarber;
+    mTheCustomers.clear();
+}
+
+void Shop::newCustomerArrives(Customer *newCustomer)
+{
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    mTheCustomers.emplace_back(newCustomer);
+}
+
+bool Shop::emptySeatsExist()
+{
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    return ((mWaitingCustomers < mMaxChairs));
+}
+
+void Shop::addWaitingCustomer()
+{
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    mWaitingCustomers++;
+}
 
+void Shop::removeWaitingCustomer()
+{
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    mWaitingCustomers--;
+    clearFirstTerminatedCustomerFound();
 }
 
-void Shop::newCustomer(std::string customerName)
+void Shop::clearFirstTerminatedCustomerFound()
 {
-    mCustomersMapMutex.lock();
-    mTheCustomers.emplace(customerName, Customer(customerName, this));
-    mCustomersMapMutex.unlock();
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    // Check if Customers have been finished
+    auto it = mTheCustomers.begin();
+    Customer* c;
+    for (;it != mTheCustomers.end(); ++it)
+    {
+        if (!((*it)->isTerminated()))
+        {
+            continue;
+        }
+        c = (*it);
+        break;
+    }
+    *it = mTheCustomers.back();
+    c->joinThread();
+    delete c;
```

Comment:
This is overly complex and I suspect buggy. You assume that immediately after the "remove waiting customer" that customer will have terminated? What if he is still getting a haircut? Also, why don't you remove him from mTheCustomers or at the very least make the pointer that points to him null? If you keep deleted customers in the list, the second time this function runs, you will ask a deleted customer if he has terminated. You will touch freed memory and boom :-(

Reply:
In order to delete an item from vector, I check first if Customer has been terminated (the member value that becomes true after). I followed a suggestion of yours regarding removing items from vectors. But I will simplify it.

Reply:
I understand what you intended to do but it is buggy. Currently the size of mCustomers does not decrease.

Reply:
I have simplified deletion, pushed a few minutes ago.